### PR TITLE
Ban #[unsafe_no_drop_flag] on zero-sized types.

### LINKED ISSUE #33237: consider removing/revising test/run-pass/zero-size-type-destructors.rs
(spawned off of [#rustc IRC discussion](https://botbot.me/mozilla/rustc/2016-04-27/?msg=64993325&page=1))

This test:

  https://github.com/rust-lang/rust/blob/master/src/test/run-pass/zero-size-type-destructors.rs

leverages `#[unsafe_no_drop_flag]`, coupled with a `Drop` impl that deliberately does not attempt to observe whether the value has already been  dropped, to count the number of actual drops are executed by the generated code.

The problem with this is that small variations on the same program exhibit semantics that probably should not be considered stable.

For example (playpen: http://is.gd/BwiZ17):

``` rust
#![feature(unsafe_no_drop_flag)]

static mut destructions : isize = 3;

pub fn foo() {
    #[unsafe_no_drop_flag]
    struct Foo(u32);

    impl Drop for Foo {
        fn drop(&mut self) {
            let p = self as *mut _;
            println!("dropping Foo(0x{:08x}) at {:?}", self.0, p);
            unsafe { destructions -= 1 };
        }
    };

    let x = Foo(1);
    println!("The location of  `&x` is at {:?}",  &x as *const _);
    let _x = [x, Foo(2), Foo(3)];
    println!("The location of `&_x` is at {:?}", &_x as *const _);
}

pub fn main() {
  foo();
  assert_eq!(unsafe { destructions }, 0);
}
```

prints:

```
The location of  `&x` is at 0x7fff1bfb0308
The location of `&_x` is at 0x7fff1bfb0258
dropping Foo(0x00000001) at 0x7fff1bfb0258
dropping Foo(0x00000002) at 0x7fff1bfb025c
dropping Foo(0x00000003) at 0x7fff1bfb0260
dropping Foo(0x1d1d1d1d) at 0x7fff1bfb0308
thread '<main>' panicked at 'assertion failed: `(left == right)` (left: `-1`, right: `0`)', <anon>:25
```

We continue to assert that there are three drops of `Foo`, since `x` is moved into `_x[0]`. But due to `unsafe_no_drop_flag`, there is no dynamic tracking of whether `x` is dropped or not, and therefore we end up executing the destructor four times. (The code above prints out the addresses of the local variables and of the `&mut self` in the destructor to make this very explicit; note that the `0x1d` bit pattern is the way we mark a value as dropped/uninitialized.)

---

I am filing this issue to point out that the test in the respository may be overly-conservative, in that it is explicitly testing the behavior of `unsafe_no_drop_flag` with code for which we do not guarantee the current behavior. 

In particular, this test is one of the things that is "blocking" MIR i.e. `-Z orbit`  in its current state, but I would argue that we do not guarantee the current behavior, at least not to this degree of specificity.

### PR DESCRIPTION
**Pull request by eddyb**
The only existing test we have relies on the simple in-place initialization behavior in old trans, e.g.:

``` rust
    // Runs Foo's Drop impl 3 times.
    let _x = [x, Foo, Foo];
```

``` rust
    // Runs Foo's Drop impl 4 times.
    let x = Foo;
    let _x = [x, Foo, Foo];
```

With MIR, there are more temporaries, and the test can't work without extra move tracking.

This is a `[breaking-change]`, but `#[unsafe_no_drop_flag]` is unstable, and only works for ZSTs in very simple cases, so I doubt there are real uses in the wild. Fixes #33237.

### CODE REVIEW

**Comment by pnkfelix** [comment]
I would think that to truly resolve #33237, we should *also* have a run-pass test that explicitly shows how one would write a similar test (on a non-zero-sized type) that properly consults the drop-flag state to decide whether to run the destructor or not.

(That, or point to an existing unit test that is similarly demonstrating how to write such code.)

**Comment by pnkfelix** [comment]
(I want to discuss this with either T-compiler or T-lang before merging.)

**Comment by pnkfelix** [comment]
@eddyb (if we do decide to land a check like this, it should be downgraded to a warning first, following something like RFC https://github.com/rust-lang/rfcs/pull/1589 )

**Comment by pnkfelix** [comment]
@eddyb (but first lets see what a crater run says)

**Comment by eddyb** [comment]
@pnkfelix Does that apply to changes in unstable features (which this is)?

**Comment by pnkfelix** [comment]
@eddyb yes I think this rule applies even for unstable features.

I can understand how that might sound odd, since someone who is using unstable features is already putting up with using the nightly channel (or a snapshot of nightly, or a locally built `rustc`).

But my thinking is that we want to make the runway for breaking changes a smooth incline, even when the breakage is solely applicable to unstable things.

----

@nikomatsakis may want to chime in further, or contradict me entirely for that matter. :) (rust-lang/rfcs#1589 doesn't draw a distinction AFAICT in terms of what kinds of features it is meant to apply to)

**Comment by drbo** [comment]
@eddyb I think part of the reason for this is that some crates require nightly Rust, as does Servo.

**Comment by eddyb** [comment]
@drbo Sure, and they deal with fallout on virtually every single *manual* update of the Rust version they support (Servo in particular has a set of custom lints which use the unstable `rustc` API).

**Comment by eddyb** [comment]
Results of crater run: https://gist.github.com/83e72ed67d7424a37a1fa8e6937b245a.
Effectively 0 regressions (and a timeout).
However, I don't care enough, given that proper non-zeroing/filling dynamic drop could land in the following weeks.

**Comment by pnkfelix** [comment]
@eddyb okay, well, r=me (just fix the tidy failures).

File: src/librustc_trans/type_of.rs [hunk 1]
```diff
@@ -231,6 +231,11 @@ pub fn in_memory_type_of<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>, t: Ty<'tcx>) ->
         return llty;
     }
 
+    // FIXME(eddyb) use the layout to actually compute the LLVM type.
```
Comment (pnkfelix) [inline src/librustc_trans/type_of.rs]: (I said "r=me" but maybe this needs to actually be addressed first?)

Comment (eddyb) [inline src/librustc_trans/type_of.rs]: That's a long term goal that is blocked on removing old trans.

**Comment by eddyb** [comment]
With non-zeroing drop here, there's no need for this.
